Support for single and unique ops test

chandrasekaranpradeep opened the pull request:
Fixes https://github.com/tenstorrent/tt-forge-fe/issues/643
Fixes https://github.com/tenstorrent/tt-forge-fe/issues/598

1) Add input node datatype and node types(i.e Activation/Parameter/Constant) as data member in Operation class
2) In _determine_node_dtype function before checking framework_dtype is N/A, add a condition to check whether the framework_dtype key is present in the node attrs because the kernel/op node doesn't have framework_dtype as key inside the node attrs.
3) Created a class method in Tensor class for creating TensorFromPytorch from shape and dtypes which will be latter used in write_pytest_function method in ForgeWritter class.
3) Modifed write_pytest_function method in ForgeWriter class in  python_codegen.py file, 
     1) Takes List of module names, framework name as str, List of Tuple and each tuple contains shapes and datatypes, List of bool values to determine whether we need call process_framework_function  or not in the pytest.
     2) All the input arguments from the write_pytest_function are tied and generated as pytest parameter.
     3)  framework_output will be of TensorFromTrace, converted the TensorFromTrace to the particular framework and then passed to the compare_with_golden function
 
Single & Unique Op test:
  Instead of passing dumping input tensor to the forge module forward function, if all the inputs for the particular op is parameter or const, we will pass the parameter/const as activation/input to the forge module forward function and in the pytest function instead generating rand values for the parameter/const as activation/input will load the actual values from the model parameter/buffers with the help of load_model_params_and_buffers which will be created by write_model_parameter_function method in ForgeWriter. For the particular op, if only the some inputs are parameter/const, we will not pass parameter/const as input to the forge module forward function.
  Created write_model_parameter_function method in ForgeWriter for loading the model parameter and buffers for above cases.
  
  Unique Op tests:
     1. Created a UniqueOperations class which will be used for extracting unique operation configuration from the Ops(i.e Dict contains key as forge op function name and Operation object as it values) and the unique  operation configuration are extracted by comparing Operation name and then will compare OperandInfo (i.e Compare Operand Node Type(i.e Activation/Parameter/Constant) and then Operand shape and then Operand dtype) and if the operands info matches, will compare operation arguments and save the operand names in the association with operation argument.
    2. After extracting unique operation configuration, for each operation, will generate a test module
    which contains forge modules for different configurations and associated pytest function.

nvukobratTT reviewed (approved):
Sending two minor comments (code comments), and a few general comments (below).

1. There are few functions that are quite huge, and will be beneficial to restructure them in the future. I know that we're currently pushing this effort to work for all models, and there are still few explorations paths that we push, so I'm okay with keeping things as is until we figure out all major requirements for this to work e2e. 
2. It'll be good to think about moving unique op generation logic into separated files, just to keep things in order. For now, let's keep this bundled as is until we get functional logic. Afterwards, let's focus on restructuring and cleaning up this code section a bit.

That said @chandrasekaranpradeep can you open two tech depth issues as assigned them to the model analysis milestone?

nvukobratTT commented:
For wider context for others, here is the output example:
- [unique_ops.zip](https://github.com/user-attachments/files/17775231/unique_ops.zip)

chandrasekaranpradeep commented:
> Sending two minor comments (code comments), and a few general comments (below).
> 
> 1. There are few functions that are quite huge, and will be beneficial to restructure them in the future. I know that we're currently pushing this effort to work for all models, and there are still few explorations paths that we push, so I'm okay with keeping things as is until we figure out all major requirements for this to work e2e.
> 2. It'll be good to think about moving unique op generation logic into separated files, just to keep things in order. For now, let's keep this bundled as is until we get functional logic. Afterwards, let's focus on restructuring and cleaning up this code section a bit.
> 
> That said @chandrasekaranpradeep can you open two tech depth issues as assigned them to the model analysis milestone?

Sure Nikola I will create two issue, one for moving unique op generation logic to separate file and another for restructure the function that are huge and assign them with model analysis milestone.

chandrasekaranpradeep commented on forge/forge/config.py:
```diff
@@ -187,6 +187,9 @@ class CompilerConfig:
     tvm_generate_op_tests: bool = False
 
+    # If enabled, for given test, it generates Forge Modules in form of PyTest for each unique behaviors that exists in given module
+    tvm_generate_unique_op_tests: bool = False
```
**tvm_generate_op_tests** will generate tests for every node in the model. **tvm_generate_unique_op_tests** will generate separate file for each operation and that contains test with unique configuration. I can't find the functional scenario right now, but if we don't need this we can remove it while restructuring the code.

nvukobratTT replied:
Sounds good. Let's see will there be a value of keeping it in the future

chandrasekaranpradeep commented on forge/forge/config.py:
```diff
@@ -187,6 +187,9 @@ class CompilerConfig:
     # If enabled, for given test, it generates Forge Modules in form of PyTest for each op that exists in given module
     tvm_generate_op_tests: bool = False
 
+    # If enabled, for given test, it generates Forge Modules in form of PyTest for each unique behaviors that exists in given module
```
Sure Nikola

chandrasekaranpradeep replied:
Updated the comment

chandrasekaranpradeep commented:
@nvukobratTT  Moved the run_tests.py inside the scripts folder and modified the run_tests function to work with both single and unique ops tests and added the ArgumentParser.

nvukobratTT commented:
> @nvukobratTT Moved the run_tests.py inside the scripts folder and modified the run_tests function to work with both single and unique ops tests and added the ArgumentParser.

Looks good! Feel free to merge the changes, and let's proceed with next issue at hand
